File: src/lib/gst.ts
export function round2(value: number): number {
  return Math.round((value + Number.EPSILON) * 100) / 100;
}

export function taxableValue(qty: number, rate: number, discount: number): number {
  return round2(qty * rate - discount);
}

export interface TaxSplit {
  cgst: number;
  sgst: number;
  igst: number;
}

export function splitTax(taxable: number, gstRate: number, intra: boolean): TaxSplit {
  if (taxable <= 0 || gstRate <= 0) return { cgst: 0, sgst: 0, igst: 0 };
  if (intra) {
    const half = round2((taxable * gstRate) / 200);
    return { cgst: half, sgst: round2(taxable * gstRate / 100 - half), igst: 0 };
  }
  return { cgst: 0, sgst: 0, igst: round2((taxable * gstRate) / 100) };
}

export function taxTotal(split: TaxSplit): number {
  return round2(split.cgst + split.sgst + split.igst);
}

File: mobile/lib/src/gst.dart
double round2(double value) =>
    ((value + 2.220446049250313e-16) * 100).round() / 100;

double taxableValue(double qty, double rate, double discount) =>
    round2(qty * rate - discount);

class TaxSplit {
  final double cgst;
  final double sgst;
  final double igst;

  const TaxSplit({this.cgst = 0, this.sgst = 0, this.igst = 0});

  double get total => round2(cgst + sgst + igst);
}

TaxSplit splitTax(double taxable, double gstRate, bool intra) {
  if (taxable <= 0 || gstRate <= 0) return const TaxSplit();
  if (intra) {
    final half = round2((taxable * gstRate) / 200);
    return TaxSplit(
      cgst: half,
      sgst: round2(taxable * gstRate / 100 - half),
    );
  }
  return TaxSplit(igst: round2((taxable * gstRate) / 100));
}